### DISCUSSION TITLE
CP-SAT Query on dynamic calculation using a Matrix

### DISCUSSION THREAD

**Discussion by SumiranLohani** [google/or-tools#2918] [CP-SAT questions]
Hello,

Here's my overall problem statement:
I have two production lines L1 and L2 which can make one or more of the following list of products:
A1, A2, B1, B2, B3, C1

The production lines are constrained by the total line capacity (input), and the productions are done to meet the production deficit (given as input).
Hence, objective - Plan production quantities for 1 or more products based on available line capacity. 

Issue:
There is a constraint related to the change of product in a line from one to another. There is time taken to changeover from one product to another determined by a matrix like below:
{(A1,A2): 5 hr, (A1,B1): 12 hrs, (A1,B2): 12 hrs, (A1,B3): 12 hrs, (A1, C1): 15 hrs,
(B1,A1): 10 hrs, (B1,A2): 12 hrs, (B1,B2): 5 hrs, (B1,B3):  5 hrs
(A2,A1): 6 hrs.....}
Hence, we have this matrix available for all product combinations based on the sequence of product change. Eg if product A1 is followed by product A2, it takes 5 hours to prepare the line for A2.
Hence, this constrains the total production capacity because time is used up in changing the product from one to another.

So I want to setup a constaint as below:
sum(production quantities of all product in a line) <= line_capacity - production_loss_due_to_product_Change

In the above equation, 
production_loss_due_to_product_Change = (total time lost due to production change) * production_per_unit_time

For eg, if we produce A1 and A2, we lose 5 hours (changeover time considering the sequence with the min total time: A1 ->A2).
production_loss_due_to_product_Change = 5 * production_per_unit_time; this gives the total quantity of products that could not be produced in the line due to changeover.

I am able to setup a test case successfully when I take total time lost due to production change = total_product_changes * constant_hrs_lost_per_product_change* 

However, I want to make it dynamic so that it is able to lookup from the matrix and calculate this. As I am new to OR Tools, I am unsure how to tackle this part, especially because this part of the problem feels like a sequencing problem, while my overall objective is tackled using a scheduling approach of CP-SAT solver.
Thanks for the support.

**Comment by lperron** [comment]
Isn't this just a TSP  with a limit on the total tour ?

**Reply by SumiranLohani** [reply]
Thanks @lperron for your response.
I was able to do a demo case with TSP and it successfully calculates the total tour (equivalent to the total changeover time).

However, the challenge I face is that this process is to be done on a week-by-week basis. Not all products are to be produced each week. So, for one week it may be optimal to produce A1, A2 and B1.
While in another week it may be optimal to produce just B1, A1, B2, C1.

So how do I update the TSP so that the nodes can be switched on and off dynamically depending on the value of a boolean : 1 - if the product is produced, consider it for the TSP and if 0 - the product is not produced and should not be considered in the TSP objective?

**Reply by lperron** [reply]
Are these different problems, solved independently?

**Reply by SumiranLohani** [reply]
Yes, every week's optimization problem is a different problem, but output of one week's optimization will affect the next week's inputs. So it is a sequential problem which I am currently tackling by looping over the no of weeks, with one week's outputs updating the inputs for the next week.

**Comment by lperron** [comment]
Then just inject the solution of last week into the current week.

Le mer. 24 nov. 2021, 07:48, SumiranLohani ***@***.***> a
écrit :

**Reply by SumiranLohani** [reply]
@lperron 
I understand the overview. 
However, heres my specific doubt:  ​
I am trying the Circuit Constraint to refer to the matrix & return the min output. My reference is - 
https://www.xiang.dev/cp-sat/
Section : Circuit constraint (ordering)
 
The objective function looks as below - 
model1.Minimize(sum(literals[i+1, j+1]*matrix_data[i+1,j+1] for i, j in permutations(all_products, 2)))
where, all_products = range(0,6)

Now, in order to prevent some products from being a part of the arc, I am trying to use booleans using the approach you have suggested in another forum: 
all_arcs.append([1,1,True])
I wanted this to happen dynamically, and hence I have provided a list of booleans corresponding to each product and appended to the all_arcs dictionary as below:
for i in all_products:
        a = production_plan[0,i].prodn_bool
        all_arcs.append([i+1,i+1,a])
Where prodn_bool is a boolean Variable for each product
So can OR Tools Boolvars be used to turn On and Off specific nodes in a circuit? 
This approach is not working for me as desired, and hence, I am wondering what's wrong with my approach.
It does work when I hardcode all_arcs.append([i+1,i+1,True]) but not with Boolvars. Any way to resolve this?

**Reply by lperron** [reply]
what does 'dynamically' means ? In the same solve ? Or when creating the model ?
